### PR TITLE
(presto-elasticsearch): migrate TransportClient to RestHighLevelClient

### PR DESCRIPTION
**Pull request by moxheit**
## Overview
This PR replaces the `TransportClient` (TC) used for communicating with Elasticsearch nodes with `RestHighLevelClient` (RHLC). TC is deemed deprecated in current (7.x) version of Elasticsearch and will be **removed** when Elasticsearch reaches 8.x.

### tl;dr
- update the client: drop 5.x, works with 6.x+
- change SSL setup due to client update
- tweak how the query gets distributed against shards

### Compatibility
- Starting at version 6.x Elasticsearch dropped concept of document types, now an index can only have one type of a document - `_doc`. This means we cannot support Elasticseach indices which are at version 5.x since they might have other types of documents than `_doc` indexed
- This PR relies on [sliced scroll API](https://www.elastic.co/guide/en/elasticsearch/reference/6.6/search-request-scroll.html#sliced-scroll) which was introduced in 5.7. This means that if we were to support 5.x we would be able to support only two versions: 5.6 and final one, 5.7.
- Upgrading to 7.x would be a matter of changing few methods, however I decided to stay at 6.x since Elasticsearch only guarantees compatibility between major version of the client and servers
- This PR breaks compatibility with Search Guard plugin. SG plugin works only with `TransportClient`. Please correct me if I'm wrong. I've implemented a bunch of helper classes to handle SSL setup, however it's not tested. ANY HELP WOULD BE HIGHLY APPRECIATED :heart:.

### Implementation
- To distribute queries we're using [sliced scroll API](https://www.elastic.co/guide/en/elasticsearch/reference/6.6/search-request-scroll.html#sliced-scroll), knowing the number of shards we can distribute parts of the scroll across all the shards. This is similar to what the plugin used to do getting shards from `TransportClient` and distributing the query. A split maps to a sliced scroll.
- SSL setup uses suggestions from Search Guard plugin's maintainer how to user SSL setup with RHLC

## Next steps & help needed
- [x] Update docs section on compatibility
- [ ] PR review suggestions
- [ ] Please help me test the SSL setup :heart:

Ps. I'm a Java noob, so any, even the tiniest feedback is welcome and will help.

### CODE REVIEW

**Comment by knikel** [comment]
@zhenxiao: This PR is ready for review and testing :pray:

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/PemReader.java [hunk 1]
```diff
@@ -0,0 +1,172 @@
+package com.facebook.presto.elasticsearch;
```
Comment (ebyhr) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/PemReader.java]: We need to add license header.

**Comment by ebyhr** [comment]
I'm looking this PR with local Elasticsearch 7.2.0 environment (default setting), but I got below error. Do you have any idea about this cause?

```sh
curl -X PUT "localhost:9200/twitter/_doc/1" -H 'Content-Type: application/json' -d'
{
    "user" : "kimchy",
    "post_date" : "2009-11-15T14:12:12",
    "message" : "trying out Elasticsearch"
}
'
```

```json
{
  "tableName": "twitter",
  "schemaName": "tpch",
  "host": "localhost",
  "port": "9200",
  "clusterName": "elasticsearch",
  "index": "twitter",
  "columns": [
    {
      "name": "user",
      "type": "varchar",
      "jsonPath": "user",
      "jsonType": "varchar",
      "ordinalPosition": "0"
    },
    {
      "name": "post_date",
      "type": "varchar",
      "jsonPath": "post_date",
      "jsonType": "varchar",
      "ordinalPosition": "1"
    },
    {
      "name": "message",
      "type": "varchar",
      "jsonPath": "message",
      "jsonType": "varchar",
      "ordinalPosition": "2"
    }
  ]
}

```

```
presto:tpch> select * from twitter;

Query 20190706_150258_00002_3hjjm, FAILED, 1 node
http://localhost:8080/ui/query.html?20190706_150258_00002_3hjjm
Splits: 17 total, 0 done (0.00%)
CPU Time: 0.0s total,     0 rows/s,     0B/s, 7% active
Per Node: 0.0 parallelism,     0 rows/s,     0B/s
Parallelism: 0.0
Peak Memory: 0B
0:05 [0 rows, 0B] [0 rows/s, 0B/s]

Query 20190706_150258_00002_3hjjm failed: java.lang.IllegalArgumentException: max must be greater than 1
java.lang.RuntimeException: java.lang.IllegalArgumentException: max must be greater than 1
	at com.facebook.presto.elasticsearch.ElasticsearchRecordCursor.getSearchResponse(ElasticsearchRecordCursor.java:195)
	at com.facebook.presto.elasticsearch.ElasticsearchRecordCursor.sendElasticsearchQuery(ElasticsearchRecordCursor.java:214)
	at com.facebook.presto.elasticsearch.ElasticsearchRecordCursor.<init>(ElasticsearchRecordCursor.java:84)
	at com.facebook.presto.elasticsearch.ElasticsearchRecordSet.cursor(ElasticsearchRecordSet.java:52)
	at com.facebook.presto.spi.RecordPageSource.<init>(RecordPageSource.java:37)
	at com.facebook.presto.split.RecordPageSourceProvider.createPageSource(RecordPageSourceProvider.java:42)
	at com.facebook.presto.split.PageSourceManager.createPageSource(PageSourceManager.java:56)
	at com.facebook.presto.operator.TableScanOperator.getOutput(TableScanOperator.java:239)
	at com.facebook.presto.operator.Driver.processInternal(Driver.java:379)
	at com.facebook.presto.operator.Driver.lambda$processFor$8(Driver.java:283)
	at com.facebook.presto.operator.Driver.tryWithLock(Driver.java:675)
	at com.facebook.presto.operator.Driver.processFor(Driver.java:276)
	at com.facebook.presto.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1077)
	at com.facebook.presto.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:162)
	at com.facebook.presto.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:483)
	at com.facebook.presto.$gen.Presto_null__testversion____20190706_145537_1.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalArgumentException: max must be greater than 1
	at org.elasticsearch.search.slice.SliceBuilder.setMax(SliceBuilder.java:165)
	at org.elasticsearch.search.slice.SliceBuilder.<init>(SliceBuilder.java:105)
	at org.elasticsearch.search.slice.SliceBuilder.<init>(SliceBuilder.java:93)
	at com.facebook.presto.elasticsearch.ElasticsearchQueryBuilder.buildScrollSearchRequest(ElasticsearchQueryBuilder.java:117)
	at com.facebook.presto.elasticsearch.ElasticsearchRecordCursor.lambda$getSearchResponse$0(ElasticsearchRecordCursor.java:192)
	at com.facebook.presto.elasticsearch.RetryDriver.run(RetryDriver.java:80)
	at com.facebook.presto.elasticsearch.ElasticsearchRecordCursor.getSearchResponse(ElasticsearchRecordCursor.java:192)
	... 18 more
	Suppressed: java.lang.IllegalArgumentException: max must be greater than 1
		... 25 more
	Suppressed: java.lang.IllegalArgumentException: max must be greater than 1
		... 25 more
	Suppressed: java.lang.IllegalArgumentException: max must be greater than 1
		... 25 more
	Suppressed: java.lang.IllegalArgumentException: max must be greater than 1
		... 25 more
```

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchClient.java [hunk 2]
```diff
@@ -198,48 +188,48 @@ public ElasticsearchTableDescription getTable(String schemaName, String tableNam
         if (tableDescription.getIndexExactMatch()) {
             return ImmutableList.of(tableDescription.getIndex());
         }
-        TransportClient client = clients.get(tableDescription.getClusterName());
+        RestHighLevelClient client = clients.get(tableDescription.getClusterName());
         verify(client != null, "client is null");
-        String[] indices = getIndices(client, new GetIndexRequest());
+        GetIndexRequest request = new GetIndexRequest();
```
Comment (ebyhr) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchClient.java]: Unused line.

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchQueryBuilder.java [hunk 3]
```diff
@@ -80,46 +83,76 @@ public ElasticsearchQueryBuilder(List<ElasticsearchColumnHandle> columnHandles,
...
         }
         catch (UnknownHostException e) {
             throw new PrestoException(ELASTICSEARCH_CONNECTION_ERROR, format("Error connecting to search node (%s:%d)", split.getSearchNode(), split.getPort()), e);
         }
-        client = createTransportClient(config, new TransportAddress(address, split.getPort()));
+        try {
+            client = createClient(config, address.getHostName(), split.getPort());
+        }
+        catch (IOException | GeneralSecurityException e) {
+            throw new PrestoException(ELASTICSEARCH_INVALID_SSL_CONFIG, format("Error while trying to set up SSL configuration"));
```
Comment (ebyhr) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchQueryBuilder.java]: Redundant function format.

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java [hunk 4]
```diff
@@ -189,9 +189,7 @@ private SearchResponse getSearchResponse(ElasticsearchQueryBuilder queryBuilder)
             return retry()
                     .maxAttempts(maxAttempts)
                     .exponentialBackoff(maxRetryTime)
-                    .run("searchRequest", () -> queryBuilder.buildScrollSearchRequest()
-                            .execute()
-                            .actionGet(requestTimeout.toMillis()));
```
Comment (ebyhr) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchRecordCursor.java]: Is `requestTimeout` no more needed? (ElasticsearchClient.java is also the same)

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchSplit.java [hunk 5]
```diff
@@ -59,18 +59,16 @@ public String getIndex()
...
-
     @JsonProperty
     public int getShard()
     {
         return shard;
     }
 
+    @JsonProperty
+    public int getShards()
+    { return shards; }
```
Comment (ebyhr) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchSplit.java]:
```suggestion
    {
        return shards;
    }
```

File: presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchSplit.java [hunk 6]
```diff
@@ -112,7 +110,6 @@ public String toString()
     {
         return toStringHelper(this)
                 .addValue(index)
-                .addValue(type)
                 .addValue(shard)
```
Comment (ebyhr) [inline presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchSplit.java]: I think `.addValue(shards)` should be added.

**Comment by knikel** [comment]
Thanks for the review @ebyhr, I'll work on this PR this week and test it against `Elasticsearch@7.x` with your scenario.

**Comment by knikel** [comment]
@ebyhr: I've tested against `Elasticsearch@7.2.0`, and the error you got shouldn't appear anymore. I didn't see this issue until you flagged it since I was working with multi shard Elasticsearch clusters from the start. In my latest change, we slice the scroll over multiple shards only when there is more than one shard.

I still have to test the SSL and rename some properties to match the latest version of SG. I'll address your other comments in the next commit. Thank you for taking the time to review it.

**Comment by moxheit** [comment]
Sorry, I won't have the capacity to finish this PR and it's been laying around untouched since July. If anyone is interested they're free to take over the little code there is in it.

**Pull request closed without merging**